File: src/code-samples/langsmith/smithdb-migration/experiment-runs-query-sort-after.py
# :remove-start:
from langsmith import Client

_setup_client = Client()
_DATASET_NAME = "docs-experiment-runs-query-fixture"
if not _setup_client.has_dataset(dataset_name=_DATASET_NAME):
    _dataset = _setup_client.create_dataset(dataset_name=_DATASET_NAME)
    _setup_client.create_examples(
        dataset_id=_dataset.id,
        examples=[
            {"inputs": {"question": "2 + 2"}, "outputs": {"answer": "4"}},
            {"inputs": {"question": "3 + 3"}, "outputs": {"answer": "6"}},
            {"inputs": {"question": "4 + 4"}, "outputs": {"answer": "9"}},
        ],
    )


def _target(inputs: dict) -> dict:
    a, b = (int(x) for x in inputs["question"].split(" + "))
    return {"answer": str(a + b)}


def correctness(outputs: dict, reference_outputs: dict) -> bool:
    return outputs["answer"] == reference_outputs["answer"]


_results = _setup_client.evaluate(
    _target,
    data=_DATASET_NAME,
    evaluators=[correctness],
    experiment_prefix="docs-experiment-runs-query-sort",
)
dataset_id = _results.get_dataset_id()
experiment_name = _results.experiment_name
# Sorting queries derive their time window from the experiment's start time,
# truncated to whole seconds. Without a short buffer, a fixture created and
# queried within the same second can produce an empty (min == max) window.
import time as _time

_time.sleep(1)
# :remove-end:

# :snippet-start: experiment-runs-query-sort-after-py
# :codegroup-tab: After
from langsmith import Client
import asyncio


async def main():
    client = Client()
    experiment_id = client.read_project(project_name=experiment_name).id
    page = await client.datasets.experiment_runs.query(
        str(dataset_id),
        experiment_ids=[str(experiment_id)],
        sort={"by": "feedback.correctness", "order": "ASC"},
    )
    return page.items


examples_with_runs = asyncio.run(main())
# :snippet-end:

# :remove-start:
if __name__ == "__main__":
    assert len(examples_with_runs) == 3
    print("✓ experiment-runs-query-sort")
# :remove-end:

File: src/code-samples/langsmith/smithdb-migration/experiment-runs-query-sort-after.ts
// :snippet-start: experiment-runs-query-sort-after-js
// :codegroup-tab: After
import { Client } from "langsmith";

const client = new Client();
// :remove-start:
const DATASET_NAME = "docs-experiment-runs-query-fixture";
const EXPERIMENT_NAME = "docs-experiment-runs-query-fixture-experiment";

if (!(await client.hasDataset({ datasetName: DATASET_NAME }))) {
  const newDataset = await client.createDataset(DATASET_NAME);
  await client.createExamples([
    { inputs: { question: "2 + 2" }, outputs: { answer: "4" }, dataset_id: newDataset.id },
    { inputs: { question: "3 + 3" }, outputs: { answer: "6" }, dataset_id: newDataset.id },
    { inputs: { question: "4 + 4" }, outputs: { answer: "9" }, dataset_id: newDataset.id },
  ]);
}
const dataset = await client.readDataset({ datasetName: DATASET_NAME });
const datasetId = dataset.id;

// The experiment is shared across every experiment-runs-query sample (this
// file and its siblings): created once, ever, and reused afterward so the
// suite doesn't spend a real evaluation run per file.
if (!(await client.hasProject({ projectName: EXPERIMENT_NAME }))) {
  await client.createProject({
    projectName: EXPERIMENT_NAME,
    referenceDatasetId: datasetId,
  });
  for await (const example of client.listExamples({ datasetId })) {
    const [a, b] = (example.inputs.question as string).split(" + ").map(Number);
    const answer = String(a + b);
    const runId = crypto.randomUUID();
    const now = new Date().toISOString();
    await client.createRun({
      id: runId,
      name: "target",
      run_type: "chain",
      inputs: example.inputs,
      outputs: { answer },
      reference_example_id: example.id,
      project_name: EXPERIMENT_NAME,
      start_time: now,
      end_time: now,
    });
    const score = answer === (example.outputs?.answer as string) ? 1 : 0;
    await client.createFeedback(runId, "correctness", { score });
  }
  // Sorting queries derive their time window from the experiment's start
  // time, truncated to whole seconds server-side. A short buffer avoids a
  // same-second min/max window on whichever run performs this creation.
  await new Promise((resolve) => setTimeout(resolve, 1000));
}
const experimentName = EXPERIMENT_NAME;
// :remove-end:
const experimentId = (await client.readProject({ projectName: experimentName })).id;
const page = await client.datasets.experimentRuns.query(datasetId, {
  experiment_ids: [experimentId],
  sort: { by: "feedback.correctness", order: "ASC" },
});
// :remove-start:
void page;
// :remove-end:
// :snippet-end:
